### PR TITLE
Migrate to uv

### LINKED ISSUE #804: Switch to uv for dependency management
### What's the issue?

uv is a faster, more reproducible package manager than pip.

Replace requirements.txt with uv.lock. Simplifies and speeds up CI.

### PR DESCRIPTION
**Pull request by rg936672**
### PR Type

- Build related changes
- CI related changes

### Description

Migrate to `uv` for dependency management. Closes #804.

### How Has This Been Tested?

See if all the workflows still pass.

### Does this PR introduce a breaking change?

Not to end users!

### Checklist before requesting a review
- [ ] I have made sure that my PR is not a duplicate.
- [ ] My code follows the style guidelines of this project.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have performed a self-review of my code.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] Any dependent changes have been merged and published in downstream modules.

### CODE REVIEW

**Comment by rg936672** [comment]
I've hidden it as outdated, but https://github.com/gchq/coreax/pull/806#issuecomment-2410702965 is a very good illustration of just how important having the latest dependencies is for performance - we experience an up-to-10x slowdown with older package versions!

File: pyproject.toml [hunk 1]
```diff
@@ -68,6 +69,14 @@ dev = [
     "ruff",
 ]
 
+# ensure we get the latest package versions available for each Python version
```
Comment (tp832944) [inline pyproject.toml:72]: Capital letter please.

File: .github/workflows/docs.yml [hunk 2]
```diff
@@ -14,40 +14,40 @@ jobs:
     runs-on: ubuntu-latest
```
Comment (tp832944) [inline .github/workflows/docs.yml:14]: Delete ` with latest dependencies` on line above.

**Review by tp832944 — CHANGES_REQUESTED** [review]

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -53,6 +53,22 @@ repos:
           (?x)^(
             examples/.+/.+.(gif|png)
           )$
+      # Keep requirements-docs is up to date
```
Comment (tp832944) [inline .pre-commit-config.yaml:56]: Undo all these changes. This could produce a lot of PRs with an unrelated change when it runs locally on every developer's computer. Let's just rely on the uv autoupdate workflow.

**Review by tp832944 — CHANGES_REQUESTED** [review]
Delete requirements.txt. We have checked that nobody internal requires it.

File: .pre-commit-config.yaml (hunk 3 above)
Comment (rg936672) [reply .pre-commit-config.yaml:56]: This will only change requirements-docs if uv.lock changes; as such, it shouldn't make any changes unless the requirements in pyproject.toml change, or unless a developer manually runs `uv lock --upgrade`. (Plus, I'd have to modify the uv autoupdate _again_ to ensure that it updates requirements-docs even if uv.lock doesn't change. I can of course go and make this change, but I think you've possibly misunderstood what the pre-commit hook does here.)

Comment (tp832944) [reply .pre-commit-config.yaml]: Thanks for the explanation.

**Review by tp832944 — APPROVED** [review]
